File: backend/src/constants/roles.ts
export const ROLES = {
  ADMIN: "ADMIN",
  PURCHASE_WAREHOUSE: "PURCHASE_WAREHOUSE",
  SLITTER_PROCESSING: "SLITTER_PROCESSING",
  PRODUCTION: "PRODUCTION",
  QC: "QC",
  DISPATCH: "DISPATCH",
  SITE_EPC: "SITE_EPC",
  MANAGEMENT: "MANAGEMENT",
} as const;

export type RoleCode = (typeof ROLES)[keyof typeof ROLES];

export const ROLE_DEFINITIONS: Array<{
  code: RoleCode;
  name: string;
  description: string;
}> = [
  {
    code: ROLES.ADMIN,
    name: "Admin",
    description:
      "Full access; manages users, roles, master data, and system configuration.",
  },
  {
    code: ROLES.PURCHASE_WAREHOUSE,
    name: "Purchase / Warehouse Team",
    description:
      "Creates Coil Master records; uploads MTC/invoice; manages storage/inspection entries.",
  },
  {
    code: ROLES.SLITTER_PROCESSING,
    name: "Slitter / Processing Team",
    description:
      "Records slitting events, slit coil generation, and dispatch to Sunrack.",
  },
  {
    code: ROLES.PRODUCTION,
    name: "Production Team",
    description:
      "Records issue of slit coils to production, production orders, batch numbers, and quantities produced.",
  },
  {
    code: ROLES.QC,
    name: "QC Team",
    description:
      "Records QC inspection results, remarks, and photos; approves/rejects batches.",
  },
  {
    code: ROLES.DISPATCH,
    name: "Dispatch Team",
    description:
      "Creates dispatch records to project sites; manages vehicle/transporter details.",
  },
  {
    code: ROLES.SITE_EPC,
    name: "Site Team / EPC Coordinator",
    description:
      "Confirms site receipt and installation details; uploads installation photos.",
  },
  {
    code: ROLES.MANAGEMENT,
    name: "Management",
    description:
      "Read-only access to all modules; views dashboards, traceability reports, and complaint analytics.",
  },
];

File: backend-go/internal/constants/roles.go
package constants

type RoleCode string

const (
	RoleAdmin             RoleCode = "ADMIN"
	RolePurchaseWarehouse RoleCode = "PURCHASE_WAREHOUSE"
	RoleSlitterProcessing RoleCode = "SLITTER_PROCESSING"
	RoleProduction        RoleCode = "PRODUCTION"
	RoleQC                RoleCode = "QC"
	RoleDispatch          RoleCode = "DISPATCH"
	RoleSiteEPC           RoleCode = "SITE_EPC"
	RoleManagement        RoleCode = "MANAGEMENT"
)

type RoleDefinition struct {
	Code        RoleCode `json:"code"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
}

var RoleDefinitions = []RoleDefinition{
	{RoleAdmin, "Admin", "Full access; manages users, roles, master data, and system configuration."},
	{RolePurchaseWarehouse, "Purchase / Warehouse Team", "Creates Coil Master records; uploads MTC/invoice; manages storage/inspection entries."},
	{RoleSlitterProcessing, "Slitter / Processing Team", "Records slitting events, slit coil generation, and dispatch to Sunrack."},
	{RoleProduction, "Production Team", "Records issue of slit coils to production, production orders, batch numbers, and quantities produced."},
	{RoleQC, "QC Team", "Records QC inspection results, remarks, and photos; approves/rejects batches."},
	{RoleDispatch, "Dispatch Team", "Creates dispatch records to project sites; manages vehicle/transporter details."},
	{RoleSiteEPC, "Site Team / EPC Coordinator", "Confirms site receipt and installation details; uploads installation photos."},
	{RoleManagement, "Management", "Read-only access to all modules; views dashboards, traceability reports, and complaint analytics."},
}

type PermissionMap map[ModuleCode]ModuleAccess

func nonePermissions() PermissionMap {
	m := make(PermissionMap, len(AllModules))
	for _, mod := range AllModules {
		m[mod] = AccessNone
	}
	return m
}

func fullPermissions() PermissionMap {
	m := make(PermissionMap, len(AllModules))
	for _, mod := range AllModules {
		m[mod] = AccessFull
	}
	return m
}

func readAllPermissions() PermissionMap {
	m := make(PermissionMap, len(AllModules))
	for _, mod := range AllModules {
		m[mod] = AccessRead
	}
	return m
}

func mergePermissions(base PermissionMap, overrides map[ModuleCode]ModuleAccess) PermissionMap {
	out := make(PermissionMap, len(base))
	for k, v := range base {
		out[k] = v
	}
	for k, v := range overrides {
		out[k] = v
	}
	return out
}

var DefaultRolePermissions = map[RoleCode]PermissionMap{
	RoleAdmin: fullPermissions(),
	RolePurchaseWarehouse: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleCoilMaster:     AccessWrite,
		ModuleSlitting:       AccessRead,
		ModuleSunrackReceipt: AccessWrite,
		ModuleDocuments:      AccessWrite,
		ModuleTraceability:   AccessRead,
		ModuleComplaint:      AccessRead,
	}),
	RoleSlitterProcessing: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleCoilMaster:   AccessRead,
		ModuleSlitting:     AccessWrite,
		ModuleDocuments:    AccessWrite,
		ModuleTraceability: AccessRead,
		ModuleComplaint:    AccessRead,
	}),
	RoleProduction: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleSlitting:       AccessRead,
		ModuleSunrackReceipt: AccessRead,
		ModuleProduction:     AccessWrite,
		ModuleFinishedGoods:  AccessWrite,
		ModuleDocuments:      AccessWrite,
		ModuleTraceability:   AccessRead,
		ModuleComplaint:      AccessRead,
	}),
	RoleQC: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleProduction:    AccessRead,
		ModuleFinishedGoods: AccessRead,
		ModuleQCInspection:  AccessWrite,
		ModuleDocuments:     AccessWrite,
		ModuleTraceability:  AccessRead,
		ModuleComplaint:     AccessRead,
	}),
	RoleDispatch: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleFinishedGoods: AccessRead,
		ModuleQCInspection:  AccessRead,
		ModuleDispatch:      AccessWrite,
		ModuleDocuments:     AccessWrite,
		ModuleTraceability:  AccessRead,
		ModuleComplaint:     AccessRead,
	}),
	RoleSiteEPC: mergePermissions(nonePermissions(), map[ModuleCode]ModuleAccess{
		ModuleDispatch:         AccessRead,
		ModuleSiteInstallation: AccessWrite,
		ModuleDocuments:        AccessWrite,
		ModuleTraceability:     AccessRead,
		ModuleComplaint:        AccessWrite,
	}),
	RoleManagement: mergePermissions(readAllPermissions(), map[ModuleCode]ModuleAccess{
		ModuleUsersRoles: AccessNone,
		ModuleDashboard:  AccessRead,
	}),
}
